cdap-app-fabric/src/main/java/io/cdap/cdap/internal/sysapp/SystemAppEnableExecutor.java

```diff
@@ -0,0 +1,152 @@
+
+
+    // TODO(CDAP-16243): Make deploy app and start programs idempotent to not have transactional issues.
+    ApplicationWithPrograms appDetail = retryableDeploySystemApp(arguments);
+    if(appDetail == null) return;
+
+    // TODO(CDAP-16243): Fix logic to restart programs for app with difference version/config.
+    for (ProgramDescriptor program : appDetail.getPrograms()) {
+      try {
+        startProgram(program.getProgramId());
+      } catch (Exception ex) {
+        LOG.error("Failed to start program {} for app {} with exception {}",
+          program.getProgramId(), arguments.getId(), ex);
+      }
+    }
+  }
+
+  private ApplicationWithPrograms retryableDeploySystemApp(Arguments arguments) {
+    try {
+      return Retries.callWithRetries(() -> {
+            try {
+              return deploySystemApp(arguments);
+            } catch (RetryableException ex) {
+              LOG.debug("Failed to deploy app {} with exception {}. Will retry", arguments.getId(), ex);
+              ex.printStackTrace();
+            }
+            return null;
+          },
+          RetryStrategies.fixDelay(6, TimeUnit.SECONDS));
+    } catch (Exception ex) {
+      LOG.error("Failed to deploy app {} with exception {}",
+          arguments.getId(), ex);
+      return null;
+    }
+  }
+
+  private ApplicationWithPrograms deploySystemApp(Arguments arguments) throws Exception {
+    ApplicationId appId = arguments.getId();
+    ArtifactSummary artifactSummary = arguments.getArtifact();
+
+    KerberosPrincipalId ownerPrincipalId =
+        arguments.getOwnerPrincipal() == null ? null : new KerberosPrincipalId(arguments.getOwnerPrincipal());
+
+    // if we don't null check, it gets serialized to "null"
+    String configString = arguments.getConfig() == null ? null : GSON.toJson(arguments.getConfig());
+
+    try {
+      return appLifecycleService
+        .deployApp(appId.getParent(), appId.getApplication(), appId.getVersion(),
+          artifactSummary, configString, x -> { }, ownerPrincipalId, arguments.canUpdateSchedules());
+    } catch (NotFoundException | UnauthorizedException | InvalidArtifactException e) {
+      // these exceptions are for sure not retry-able. It's hard to tell if the others are, so we just try retrying
+      // up to the default time limit
+      throw e;
+    } catch (DatasetManagementException e) {
+      if (e.getCause() instanceof UnauthorizedException) {
+        throw (UnauthorizedException) e.getCause();
+      } else {
+        throw new RetryableException(e);
+      }
+    } catch (Exception e) {
+      throw new RetryableException(e);
+    }
+  }
+
+  private void startProgram(ProgramId programId) throws Exception {
+    Preconditions.checkArgument(programLifecycleService.getProgramSpecification(programId) != null,
+        "Cannot start %s because it does not exist.", programId);
+
+    try {
+      // do nothing if the program is already running
```

Comment:
please insert a TODO to make this restart the program if the application was changed.

Reply:
Had added similar TODO above but also added here to keep track of it.